### PR TITLE
Feat: Global Versioning

### PR DESCRIPTION
**Pull request by Bram1903**
# PacketEvents Global Versioning

This PR introduces a global versioning system for the entire library, aiming to simplify version management by removing the need to change versions across multiple files and locations.

## PEVersion
The `PEVersion` class has been completely rewritten to improve its implementation and support for the `SNAPSHOT` tag. The new class implements `Comparable<PEVersion>` and incorporates proper versioning logic. The previous integer array has been replaced with four fields: `int major`, `int minor`, `int patch`, and `boolean snapshot`.

### Key Changes:
- **Deprecated Constructor:** The old constructor is marked as deprecated.
- **New Constructors:** Two new constructors have been added. One requires semantic versioning, and the other is an overloaded method defaulting `snapshot` to false. These constructors throw an `IllegalArgumentException` if the provided version isn't in semantic versioning format.
- **SNAPSHOT Handling:** The `-SNAPSHOT` suffix is optional and defaults to false if not present.
- **Javadoc Updates:** The existing Javadoc has been revised to reflect the new implementation, with links to relevant classes and deprecated methods pointing to their replacements.
- **Comparable Implementation:** The class now implements `Comparable<PEVersion>`, enhancing readability and understanding of the version comparison logic.

### Versioning Example
- `2.3.1-SNAPSHOT` < `3.2.0`
- `2.3.1` > `2.3.1-SNAPSHOT`

`SNAPSHOT` is treated as a pre-release, thus `2.3.1` is considered newer than `2.3.1-SNAPSHOT`.

## Resource Processor
To streamline updating platform-specific version files (`plugin.yml`, `bungee.yml`, etc.), a task has been added to the `packetevents.library-convention`. This task replaces the version variable in these files during compilation, removing the need for manual updates.

### Velocity Implementation:
- The previous `@Plugin` annotation, which generated a `velocity-plugin.json` at runtime, has been removed.
- The `velocity-plugin.json` has now been manually placed in the resources directory with the version variable to be overwritten during the build process.

## Build Source PEVersion Task / Plugin
Initially, we considered adding the project version to the JAR manifest, but this approach proved unsuitable when shading the library into other plugins, as the manifest would be lost.

### Solution:
- A version file is created within the final build to indicate the library's version, ensuring compatibility even when the library is shaded.
- A new task, `PEVersionTask`, has been implemented by @AbhigyaKrishna. This task, named `generateVersionsFile`, creates a `PEVersions` file during compilation, instantiating the `PEVersion` class with the provided version.

This global versioning system simplifies version management and ensures consistent versioning across the entire library.

### CODE REVIEW

File: buildSrc/src/main/kotlin/packetevents.library-conventions.gradle.kts [hunk 1]
```diff
@@ -30,6 +30,12 @@ tasks {
         options.release = 8
     }
 
+    processResources {
+        filesMatching(listOf("plugin.yml", "bungee.yml", "velocity-plugin.json", "fabric.mod.json")) {
```
Comment (booky10) [inline buildSrc/src/main/kotlin/packetevents.library-conventions.gradle.kts:34]: The inputs should be set here to make gradle not use the cache when just changing the version

File: api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java [hunk 2]
```diff
@@ -35,11 +35,16 @@ public abstract class PacketEventsAPI<T> {
...
-    private static final PEVersion VERSION = new PEVersion(2, 3, 1);
-    
+    private final Logger LOGGER = Logger.getLogger(PacketEventsAPI.class.getName());
+
+    private static final String version = PacketEvents.class.getPackage().getImplementationVersion();
+
+    // The version needs to be checked for null, because when running Unit Tests, the version will be null.
+    private static final String[] versionParts = (version != null) ? version.split("-") : new String[]{ "0.0.0" };
+    private static final String[] parts = versionParts[0].split("\\.");
+    private static final PEVersion VERSION = new PEVersion(Integer.parseInt(parts[0]), Integer.parseInt(parts[1]), Integer.parseInt(parts[2]));
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java:46]: Imo this should be implemented as a static method in the PEVersion class

File: api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java [hunk 3]
```diff
@@ -35,11 +35,16 @@ public abstract class PacketEventsAPI<T> {
     private static final EventManager EVENT_MANAGER = new EventManager();
     private static final PacketEventsSettings SETTINGS = new PacketEventsSettings();
     private static final UpdateChecker UPDATE_CHECKER = new UpdateChecker();
-    private final Logger LOGGER = Logger.getLogger(PacketEventsAPI.class.getName());
     private static final LogManager LOG_MANAGER = new LogManager();
-    //TODO UPDATE
-    private static final PEVersion VERSION = new PEVersion(2, 3, 1);
-    
+    private final Logger LOGGER = Logger.getLogger(PacketEventsAPI.class.getName());
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java:39]: Is there a reason to this being moved down one line?

File: api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java (hunk 2 above)
Comment (Bram1903) [reply api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java:46]: Will do, thanks for the feedback!

File: buildSrc/src/main/kotlin/packetevents.library-conventions.gradle.kts (hunk 1 above)
Comment (Bram1903) [reply buildSrc/src/main/kotlin/packetevents.library-conventions.gradle.kts:34]: I am not entirely sure what you mean here? Could you elaborate?

File: api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java (hunk 2 above)
Comment (Bram1903) [reply api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java]: This has now been implemented. Can you take a look at the latest commit?

**Comment by AbhigyaKrishna** [comment]
You could also rework the `PEVersion` class. As we are always following the `major.minor.patch-SNAPSHOT` semantic, the class could contain there integers and a boolean for snapshot. And also can implement `Comparable<PEVersion>` to make it consistent.

File: buildSrc/src/main/kotlin/packetevents.library-conventions.gradle.kts (hunk 1 above)
Comment (booky10) [reply buildSrc/src/main/kotlin/packetevents.library-conventions.gradle.kts:34]: just so `inputs.property("version", project.version)` before this line

File: api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java (hunk 3 above)
Comment (booky10) [reply api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java:40]: oh nvm, i assumed it was static because of the naming

File: api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java (hunk 2 above)
Comment (booky10) [reply api/src/main/java/com/github/retrooper/packetevents/PacketEventsAPI.java]: looks good 👍

**Comment by Bram1903** [comment]
> You could also rework the `PEVersion` class. As we are always following the `major.minor.patch-SNAPSHOT` semantic, the class could contain there integers and a boolean for snapshot. And also can implement `Comparable<PEVersion>` to make it consistent.

Sure, thanks for the tip! Will do!

File: buildSrc/src/main/kotlin/packetevents.library-conventions.gradle.kts (hunk 1 above)
Comment (Bram1903) [reply buildSrc/src/main/kotlin/packetevents.library-conventions.gradle.kts:36]: Should be good now, right?

**Comment by Bram1903** [comment]
Hey @AbhigyaKrishna,

I just finished rewriting the `PEVersion` class. Could you take a look at it? It now implements both `Comparable` and `Cloneable`.
Personally, I think this is a much-improved version of the previous implementation. 

It now also has proper snapshot support, and I think it's overall more useable.

**Review by AbhigyaKrishna — CHANGES_REQUESTED** [review]

**Review by AbhigyaKrishna — APPROVED** [review]
Looks good

File: api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java [hunk 4]
```diff
@@ -18,142 +18,201 @@
...
     /**
-     * Specify your version using an array.
+     * Constructs a {@link PEVersion} instance with snapshot defaulted to false.
      *
-     * @param version Array version.
+     * @param major the major version number.
+     * @param minor the minor version number.
+     * @param patch the patch version number.
      */
-    public PEVersion(final int... version) {
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:42]: this constructor should be re-added, otherwise this would be a breaking change

Comment (Bram1903) [reply api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:42]: > this constructor should be re-added, otherwise this would be a breaking change

Shall I mark it as deprecated, and then make it use the overload constructor which defaults snapshot to false?

Comment (booky10) [reply api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:42]: Oops, missed your response - yeah, imo thats a good solution

File: api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java [hunk 5]
```diff
@@ -18,142 +18,212 @@
...
     }
 
     /**
-     * Represented as an array.
+     * Checks if the version is a snapshot.
      *
-     * @return Array version.
+     * @return true if snapshot, false otherwise.
      */
-    public int[] asArray() {
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:114]: this method has been removed, is this an intended breaking change?

Comment (Bram1903) [reply api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:114]: Yes, I wasn't sure how to properly implement it without just skipping the snapshot part. Do you want me to add it, and mark it as deprecated? This would mean snapshot won't be accounted for in the return value of that method.

Comment (booky10) [reply api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:114]: :+1: snapshot wasn't present there previously either, an `@Deprecated` is definetely needed

Comment (Bram1903) [reply api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:114]: I'll add this tomorrow morning, and test this while shading PE as well!

Comment (booky10) [reply api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:114]: You've called it "toArray", it should be named "asArray" 😄

**Comment by AbhigyaKrishna** [comment]
We are now storing the current version as a constant. This file is generated by a gradle task is generated every build.

File: api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java (hunk 5 above)
Comment (Bram1903) [reply api/src/main/java/com/github/retrooper/packetevents/util/PEVersion.java:114]: Fixed 😅

**Comment by Bram1903** [comment]
LGTM

**Review by booky10 — APPROVED** [review]
:+1:
